src/coreclr/vm/i386/asmhelpers.asm

```diff
@@ -915,6 +915,23 @@ _ProfileTailcallNaked@4 proc public
     retn    4
 _ProfileTailcallNaked@4 endp
 
+_SinglecastDelegateInvokeStub@0 proc public
+
+    test    ecx, ecx
+    jz      NullObject
+
+
+    mov     eax, [ecx + DelegateObject___methodPtr]
+    mov     ecx, [ecx + DelegateObject___target]  ; replace "this" pointer
+
+    jmp     eax
+
+NullObject:
+    mov     ecx, CORINFO_NullReferenceException_ASM
+    jmp     @JIT_InternalThrow@4
```

Comment:
This does not seem to be doing the right thing wrt. popping the arguments from the stack. I am wondering whether it is a pre-existing bug or whether the existing `JIT_InternalThrowFromHelper` has some trick in it that makes it work.

Reply:
All other platforms are already doing so. The only difference between `JIT_InternalThrow` and `JIT_InternalThrowFromHelper` is the level of stack depth (`Frame::FRAME_ATTR_EXACT_DEPTH` versus `Frame::FRAME_ATTR_CAPTURE_DEPTH_2|Frame::FRAME_ATTR_EXACT_DEPTH`).
Since the `call` is replaced with a `jmp`, the stack depth should be reduced by 1.

Reply:
All other platforms are caller-cleanup.

Windows x86 is the only callee-cleanup platform that creates all sorts of interesting problems.

Reply:
Did a simple test locally, NRE can be successfully caught under a complicated signature.
Adding a test covering this in CI should be fine?

Reply:
My guess is that we have a test for general null reference exception thrown by this helper. We may be missing the specific test that hits the potential bug that I have pointed out. Could you please share your test?

Could you please share your test?

Reply:
Just:

```csharp
            try
            {
                Func<int, string, Guid, object, double> f = null;

                double v = f(123456, "abc", Guid.NewGuid(), null);

                Console.WriteLine(v);
            }
            catch (NullReferenceException)
            {
                Console.WriteLine("NRE caught");
            }
```

It should be equivalent to setting a HMF and throw from any signature.

Reply:
Some more thoughts: it may be simpler if the invoke stub is considered as managed code, and go through the managed AV handler. However, we are not able to express the constraint to not touch any non-scratch register and must use tail jmp in IL. Replicating all the args like multicast stub may be verbose and may introduce other issues.

Reply:
This test does not exercise the null reference throwing exception in this stub.

Reply:
I'm now confused about whether the null check is reachable at all.

A plain delegate invocation is expanded by JIT so it won't hit the stub:

```csharp
    public static int X(Func<int, int> f)
    {
        return f(123) + 456;
    }
```

```asm
C.X(System.Func`2<Int32,Int32>)
    L0000: sub rsp, 0x28
    L0004: mov rax, rcx ; rcx = rax = Delegate object
    L0007: mov edx, 0x7b
    L000c: mov rcx, [rax+8] ; rcx is _target ; throws NRE(managed AV) when rax is null
    L0010: call qword ptr [rax+0x18] ; call _methodPtr
    L0013: add eax, 0x1c8
    L0018: add rsp, 0x28
    L001c: ret
```

Reflection invocation of the `Invoke` method goes through the stub (confirmed by putting something really bad in the stub), but reflection stack already blocks you from passing `null` reference to instance method `Invoke`. Can other interop case go through it? If not, we may remove the null check at all.

Reply:
OK there is a viable reproduction with another level of indirection:

```csharp
            try
            {
                Func<int, string, Guid, object, double> f = new C().M;
                Func<int, string, Guid, object, double> f2 = f.Invoke;

                var targetField = typeof(Delegate).GetField("_target", BindingFlags.NonPublic | BindingFlags.Instance);
                targetField.SetValue(f2, null);

                double v = f2(123456, "abc", Guid.NewGuid(), null);

                Console.WriteLine(v);
            }
            catch (NullReferenceException)
            {
                Console.WriteLine("NRE caught");
            }
```

The NRE is successfully caught on 8.0 x64, and also caught on this PR. Maybe the inconsistency is in reflection stack instead...

Reply:
> the exception won't be caught, but successfully prints stack trace as an unhandled exception. 

This is one of the common symptoms of messed up stack unwinind.

Given that this is rarely used method (JIT will inline it for the common case, so you will only hit it with cases like reflection that should be very rare), it may be best to turn it into IL stub. It would delete a bunch of platform specific code and the stackunwinding would be correct by construction. Note that you do not need an explicit null check in the IL stub - it is fine to depend on an implicit nullref exception to be thrown as side-effect of accessing the delegate field.

Reply:
I wonder if WrapperDelegate can be removed, since the managed IL stub will do everything correctly.

I'll add tests and JIT results tomorrow.

Reply:
Code gen for some different delegates:

```asm
; Assembly listing for method System.Action:IL_STUB_SinglecastDelegate_Invoke():this (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; optimized using Synthesized PGO
; rsp based frame
; fully interruptible
; with Synthesized PGO: fgCalledCount is 100
; No PGO data

G_M000_IG01:                ;; offset=0x0000

G_M000_IG02:                ;; offset=0x0000
       mov      rax, qword ptr [rcx+0x18]
       mov      rcx, gword ptr [rcx+0x08]

G_M000_IG03:                ;; offset=0x0008
       tail.jmp rax

; Total bytes of code 11

; Assembly listing for method System.Func`6[int,System.__Canon,int,System.__Canon,System.__Canon,System.__Canon]:IL_STUB_SinglecastDelegate_Invoke(int,System.__Canon,int,System.__Canon,System.__Canon):System.__Canon:this (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; optimized using Synthesized PGO
; rsp based frame
; fully interruptible
; with Synthesized PGO: fgCalledCount is 100
; No PGO data

G_M000_IG01:                ;; offset=0x0000

G_M000_IG02:                ;; offset=0x0000
       mov      rax, qword ptr [rcx+0x18]
       nop

G_M000_IG03:                ;; offset=0x0005
       mov      r10, gword ptr [rsp+0x28]
       mov      gword ptr [rsp+0x28], r10
       mov      r10, gword ptr [rsp+0x30]
       mov      gword ptr [rsp+0x30], r10
       mov      rcx, gword ptr [rcx+0x08]

G_M000_IG04:                ;; offset=0x001D
       tail.jmp rax

; Total bytes of code 32

; Assembly listing for method System.Func`5[int,double,System.__Canon,System.Guid,double]:IL_STUB_SinglecastDelegate_Invoke(int,double,System.__Canon,System.Guid):double:this (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; optimized using Synthesized PGO
; rsp based frame
; fully interruptible
; with Synthesized PGO: fgCalledCount is 100
; No PGO data

G_M000_IG01:                ;; offset=0x0000
       mov      rax, bword ptr [rsp+0x28]

G_M000_IG02:                ;; offset=0x0005
       mov      r8, qword ptr [rcx+0x18]
       nop

G_M000_IG03:                ;; offset=0x000A
       mov      bword ptr [rsp+0x28], rax
       mov      rcx, gword ptr [rcx+0x08]

G_M000_IG04:                ;; offset=0x0013
       tail.jmp r8

; Total bytes of code 22
```

Since it's not a hot path, it should be fine.
IL is not sharable between different delegates, since the GC tracking/stack walking information would be different.

Reply:
Deleting `WrapperDelegate ` would require changing Arm32 calling convention for VSD stubs to avoid passing the cell address in R4. It should be doable (Native AOT is on this plan already), but it should be a separate change.

Reply:
#105031